nav2_map_server/src/map_3d/map_io_3d.cpp

```diff
@@ -0,0 +1,354 @@
+/// @throw YAML::Exception
+template<typename T>
+T yaml_get_value(const YAML::Node & node, const std::string & key)
+{
+  try {
+    return node[key].as<T>();
+  } catch (YAML::Exception & e) {
+    std::stringstream ss;
+    ss << "Failed to parse YAML tag '" << key << "' for reason: " << e.msg;
+    throw YAML::Exception(e.mark, ss.str());
+  }
+}
+
+LoadParameters loadMapYaml(const std::string & yaml_filename)
+{
+  std::cout << "loading the yaml file";
+  std::string yaml_file_local = yaml_filename;
+
+  if (yaml_file_local[0] != '/') {
+    // dirname takes a mutable char *, so we copy into a vector
+    yaml_file_local =
+      std::string(SRC_DIR) + '/' + yaml_file_local;
+  }
+
+  YAML::Node doc = YAML::LoadFile(yaml_file_local);
+  LoadParameters load_parameters;
+
+  std::string pcd_file_name(yaml_get_value<std::string>(doc, "pcd"));
+  if (!ends_with(pcd_file_name, ".pcd")) {
+    throw YAML::Exception(
+            doc["image"].Mark(),
+            "The pcd file is invalid please pass a file name with extension .pcd");
+  }
+
+  if (pcd_file_name[0] != '/') {
+    // dirname takes a mutable char *, so we copy into a vector
+    load_parameters.pcd_file_name =
+      std::string(SRC_DIR) + '/' + pcd_file_name;
+  } else {
+    load_parameters.pcd_file_name = pcd_file_name;
+  }
+
+  // Get view point as position and orientation.
+  // If not provided by YAML, it will be loaded from PCD file while
+  // reading later in loadMapFromFile().
+  tf2::Quaternion rotation(0.0, 0.0, 0.0, 0.0);
+  tf2::Vector3 translation(1.0, 0.0, 0.0);
+
+  try {
+    // Load origin form yaml_file
+    std::vector<double> position = yaml_get_value<std::vector<double>>(doc, "position");
+    if (position.size() != 3) {
+      // throw std::invalid_argument("Position size should be : 3");
+      std::cout << "[WARNING] [map_io_3d]: Wrong size potision feld is provided in yaml, "
+        "will try to load origin from .pcd file" << std::endl;
+      load_parameters.position_from_pcd = true;
+    } else {
+      // Position
+      translation = tf2::Vector3(
+        tf2Scalar(position[0]),
+        tf2Scalar(position[1]),
+        tf2Scalar(position[2]));
+    }
+  } catch (YAML::Exception & e) {
+    std::cout << "[WARNING] [map_io_3d]: Couldn't load position from yaml file" << std::endl;
+  }
+
+  try {
+    std::vector<double> orientation = yaml_get_value<std::vector<double>>(doc, "orientation");
+    if (orientation.size() != 4) {
+      // throw std::invalid_argument("Position size should be : 4");
+      std::cout << "[WARNING] [map_io_3d]: Wrong size orientation feld is provided in yaml, "
+        "will try to load orientation from .pcd file" << std::endl;
+      load_parameters.orientation_from_pcd = true;
+    } else {
+      // Orientation
+      rotation = tf2::Quaternion(
+        tf2Scalar(orientation[0]),
+        tf2Scalar(orientation[1]),
+        tf2Scalar(orientation[2]),
+        tf2Scalar(orientation[3]));
+    }
```

Comment:
Why did you split `origin` -> to `position` and `orientation`? As far as I remember, in previous version of this file it was one `origin` vector and it was approved. Please return it back to `origin` for unification with 2D YAML format.